### PR TITLE
Transfer data over UDP

### PR DESCRIPTION
**Pull request by Nordsoft91**
PR is needed to test have builds for all platforms

- [ ] Fix compilation for all platforms
- [ ] Manage disconnection
- [ ] Manage local multiplayer
- [ ] Manage in-game reconnections
- [ ] Reimplement online game

### CODE REVIEW

**Comment by IvanSavenko** [comment]
Thanks for attempting to do this change. Hopefully enet is as good as described.

I know that this is very, very WIP, but perhaps create abstraction layer over enet instead of spreading it all over the code as we have with boost::asio right now? e.g:
- separate CConnection into IConnection interface + EnetConnection implementation class
- extract from CServerHandler class INetworkClient (with ENetClient as implementation)
- extract from CVCMIServerclass INetworkServer (with ENetServer as implementation)

Also potential problem with MXE (surprise). It seems that they don't provide precompiled package with enet. Meaning, we can't easily upgrade our MXE CI.

**Comment by kambala-decapitator** [comment]
> Also potential problem with MXE (surprise). It seems that they don't provide precompiled package with enet. Meaning, we can't easily upgrade our MXE CI.

one more reason to switch to mingw / msvc and use vcpkg / [conan](https://conan.io/center/enet) (both have it)

**Comment by dydzio0614** [comment]
BTW... @Nordsoft91 did you perform some early benchmark of this approach?

And side note on MXE stuff... I think MSVC is better pick for daily builds IMO for next releases anyway, since it better tuned for running resulting build on Windows in modern times while MXE's GCC 5 isn't. I would probably trust VS 2019 and 2022 optimizations more than what MXE can offer.

**Comment by IvanSavenko** [comment]
@dydzio0614 
As somebody who worked on competitive FPS game: don't expect changes to *performance*. The only change would be decrease in *latency* when playing over network with non-zero ping (read - Internet). For example huge latency from #1133 should be reduced to ping, so ~50 ms if both players are in Europe. And any overhead on top of ping can be looked into and fixed which is not possible with TCP.

I do know of network emulator tool (clumsy) that can be used to check gameplay over Internet with like 0 setup time - we used it back in our FPS, but that thing is Windows-only.

on MXE: agreed. Keep MXE for any 1.1.x releases, but for 1.2 we should use msvc & remove MXE. If somebody can setup mingw to replace MXE - great. But that is optional IMO.

**Comment by rilian-la-te** [comment]
@IvanSavenko Actually, Conan with mingw can be used easily and better than MXE. And then we can drop CMake 3.10.

**Comment by dydzio0614** [comment]
I didn't exactly mean performance but reduction of networking overhead where it is known to cause problems

**Comment by IvanSavenko** [comment]
> @IvanSavenko Actually, Conan with mingw can be used easily and better than MXE. And then we can drop CMake 3.10.

As I said, I have no objections against mingw, so if somebody can set it up - great. Not familiar with mingw or conan myself. 
But I do think that msvc should be the version we use for public releases.

BTW - we can already drop cmake 3.10, I had to upgrade MXE to include Qt translator, so MXE in develop now uses cmake 3.19. I think right now cmake 3.16 is our minimum - due to ubuntu-20.04.

**Comment by Nordsoft91** [comment]
Before making refactoring or significant architecture changes, I want to make a prototype with minimal possible efforts to test whether it actually solves problem with latency. 
So current plan is to make MVP with multiplayer

**Comment by IvanSavenko** [comment]
@Nordsoft91 If you still have latency issues with UDP, let me know - I should be able to investigate their source.

File: CI/linux/before_install.sh [hunk 1]
```diff
@@ -8,4 +8,4 @@ sudo apt-get install libsdl2-dev libsdl2-image-dev libsdl2-mixer-dev libsdl2-ttf
 sudo apt-get install qtbase5-dev
 sudo apt-get install ninja-build zlib1g-dev libavformat-dev libswscale-dev libtbb-dev libluajit-5.1-dev
 # Optional dependencies
-sudo apt-get install libminizip-dev libfuzzylite-dev qttools5-dev
+sudo apt-get install libminizip-dev libfuzzylite-dev qttools5-dev enet
```
Comment (IvanSavenko) [inline CI/linux/before_install.sh:11]:
```suggestion
sudo apt-get install libminizip-dev libfuzzylite-dev qttools5-dev libenet-dev
```

**Comment by kambala-decapitator** [comment]
@Nordsoft91 conan deps now have enet. Please apply the following patch:

```diff
diff --git a/conanfile.py b/conanfile.py
index a3030c689..1b8dcd41f 100644
--- a/conanfile.py
+++ b/conanfile.py
@@ -10,6 +10,7 @@ class VCMI(ConanFile):
     settings = "os", "compiler", "build_type", "arch"
     requires = [
         "boost/[^1.69]",
+        "enet/[~1.3.17]",
         "minizip/[~1.2.12]",
         "onetbb/[^2021.3]", # Nullkiller AI
         "qt/[~5.15.2]", # launcher
@@ -31,6 +32,7 @@ class VCMI(ConanFile):
         "with_luajit": False,
 
         "boost/*:shared": True,
+        "enet/*:shared": True,
         "minizip/*:shared": True,
         "onetbb/*:shared": True,
     }
```

**Comment by kambala-decapitator** [comment]
@Nordsoft91 please apply the patch exactly as I posted. Also please keep alphabetical order of the deps (like in my patch).

**Pull request closed without merging**